File: frontend/src/stores/logs.ts
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import { useInfiniteCursor } from '@/composables/useInfiniteCursor'
import { apiService } from '@/services/apiService'
import type { LogEntry } from '@/bindings/LogEntry'
import type { LogSourcesResponse } from '@/bindings/LogSourcesResponse'
import type { LogSourceEntry } from '@/bindings/LogSourceEntry'
import type { LogNatDiagnostics } from '@/bindings/LogNatDiagnostics'
import type { ProvisionHintsResponse } from '@/bindings/ProvisionHintsResponse'
import type { ProvisionLoggingResponse } from '@/bindings/ProvisionLoggingResponse'
import type { SetupGuide } from '@/bindings/SetupGuide'

export type {
  LogEntry,
  LogSourceEntry,
  LogNatDiagnostics,
  ProvisionHintsResponse,
  ProvisionLoggingResponse,
  SetupGuide,
}

export interface LogFilters {
  deviceId: number | null
  /**
   * Severidade numérica **máxima**. No syslog o número menor é o mais grave,
   * então `3` significa "erro e acima". `null` não filtra.
   */
  severity: number | null
  /** Janela em horas contadas para trás. `null` usa o padrão do backend (24 h). */
  hours: number | null
  search: string
}

/**
 * Opções do seletor de severidade, do mais grave para o menos.
 *
 * Os rótulos individuais vêm do backend em `severityLabel` — estes aqui
 * descrevem faixas ("erro e acima"), que é outra coisa e só existe na tela.
 */
export const SEVERITY_OPTIONS = [
  { value: 2, label: 'Crítico e acima' },
  { value: 3, label: 'Erro e acima' },
  { value: 4, label: 'Aviso e acima' },
  { value: 6, label: 'Informação e acima' },
  { value: 7, label: 'Tudo, inclusive depuração' },
] as const

/** Janelas oferecidas na tela. O backend recusa qualquer coisa além de 7 dias. */
export const WINDOW_OPTIONS = [
  { value: 1, label: 'Última hora' },
  { value: 6, label: 'Últimas 6 horas' },
  { value: 24, label: 'Últimas 24 horas' },
  { value: 24 * 7, label: 'Últimos 7 dias' },
] as const

export function defaultFilters(): LogFilters {
  return { deviceId: null, severity: null, hours: 24, search: '' }
}

export const useLogsStore = defineStore('logs', () => {
  const filters = ref<LogFilters>(defaultFilters())

  /**
   * O caminho é derivado dos filtros a cada chamada, e não guardado: é assim
   * que a mesma lista segue o filtro que o usuário acabou de mudar sem ser
   * recriada.
   */
  function endpoint(): string {
    const params = new URLSearchParams()
    if (filters.value.deviceId !== null) params.set('deviceId', String(filters.value.deviceId))
    if (filters.value.severity !== null) params.set('severity', String(filters.value.severity))
    if (filters.value.hours !== null) {
      const from = new Date(Date.now() - filters.value.hours * 3_600_000)
      params.set('from', from.toISOString())
    }
    const termo = filters.value.search.trim()
    if (termo) params.set('q', termo)
    const query = params.toString()
    return query ? `/logs?${query}` : '/logs'
  }

  const list = useInfiniteCursor<LogEntry>(endpoint, { label: 'os registros de log' })

  const isEmpty = computed(() => list.items.value.length === 0)

  /** Reinicia a lista. Chamado sempre que um filtro muda. */
  function applyFilters(next: Partial<LogFilters> = {}): void {
    filters.value = { ...filters.value, ...next }
    list.reset()
  }

  function clearFilters(): void {
    filters.value = defaultFilters()
    list.reset()
  }

  // --- live tail ------------------------------------------------------------

  const tailing = ref(false)
  let tailSource: EventSource | null = null

  /**
   * Liga o live tail.
   *
   * Stream **próprio**, não o `/api/events/stream` do painel: o barramento de
   * log é separado justamente para que um tail atrasado não derrube eventos de
   * domínio (ver `services/syslog/bus.rs` no backend).
   */
  function startTail(): void {
    if (tailSource) return
    const params = new URLSearchParams()
    if (filters.value.deviceId !== null) params.set('deviceId', String(filters.value.deviceId))
    if (filters.value.severity !== null) params.set('severity', String(filters.value.severity))
    const token = localStorage.getItem('auth_token')
    if (token) params.set('token', token)

    tailSource = new EventSource(`/api/logs/stream?${params.toString()}`)
    tailSource.onmessage = (msg) => {
      try {
        const bruto = JSON.parse(msg.data) as Record<string, unknown>
        // O primeiro quadro é só o handshake com o `retry`.
        if (bruto.type === 'stream:connected') return
        list.prepend([bruto as unknown as LogEntry], (entry) => entry.id)
      } catch {
        // Quadro ilegível não pode derrubar o tail.
      }
    }
    tailSource.onerror = () => {
      // O `EventSource` reconecta sozinho; a bandeira só reflete o estado.
      tailing.value = tailSource?.readyState !== EventSource.CLOSED
    }
    tailing.value = true
  }

  function stopTail(): void {
    tailSource?.close()
    tailSource = null
    tailing.value = false
  }

  function toggleTail(): void {
    if (tailing.value) stopTail()
    else startTail()
  }

  // --- fontes vistas --------------------------------------------------------

  const sources = ref<LogSourceEntry[]>([])
  const unknownCount = ref(0)
  const sourcesLoaded = ref(false)
  const nat = ref<LogNatDiagnostics | null>(null)

  async function fetchSources(): Promise<boolean> {
    try {
      const response = await apiService.get<LogSourcesResponse>('/logs/sources')
      sources.value = response.data ?? []
      unknownCount.value = response.unknownCount ?? 0
      nat.value = response.nat ?? null
      sourcesLoaded.value = true
      return true
    } catch {
      // A ingestão pode estar desligada; a tela de logs continua útil.
      sourcesLoaded.value = true
      return false
    }
  }

  /**
   * Se o mascaramento do Docker ainda está **atrapalhando**.
   *
   * Não é "existe mascaramento": é "existe origem mascarada sem vínculo". Os
   * roteadores chegam todos com o mesmo IP e só o hostname os separa — o aviso
   * existe para o operador não tentar resolver isso vinculando o endereço do
   * gateway, que atribuiria o parque inteiro a um dispositivo só.
   *
   * Vinculados todos, o mascaramento continua lá e deixou de custar alguma
   * coisa. Manter o aviso ali seria ensinar a ignorá-lo — e no dia em que um
   * equipamento novo aparecer sem vínculo, ele volta e volta significando algo.
   */
  const natMasking = computed(() => (nat.value?.unresolvedMaskedCount ?? 0) > 0)

  /**
   * `key` é o `bindKey` da origem, não o IP: atrás de NAT ele é `host:<nome>`.
   * Montar a chave na tela duplicaria a decisão de "isto está mascarado?", que
   * só o servidor sabe tomar.
   */
  /**
   * Vincula (ou desvincula) uma origem a um dispositivo.
   *
   * **Propaga o erro** em vez de devolver `false`: o servidor recusa o vínculo
   * do gateway do NAT com uma explicação inteira de por quê e do que fazer no
   * lugar, e engolir isso deixava o seletor voltando a vazio sem uma palavra —
   * exatamente o desfecho que a mensagem existe para evitar.
   *
   * A lista de logs **não** é recarregada: o vínculo vale para o que chegar
   * daqui em diante, e nenhuma linha já gravada muda. Recarregá-la só piscava a
   * tela atrás do diálogo.
   */
  async function bindSource(key: string, deviceId: number | null): Promise<void> {
    await apiService.post(`/logs/sources/${encodeURIComponent(key)}/bind`, { deviceId })
    await fetchSources()
  }

  // --- ativação automática --------------------------------------------------

  /**
   * O que o servidor consegue descobrir sozinho sobre o equipamento.
   *
   * Sonda as portas de acesso e consulta o SNMP, então demora alguns segundos
   * no pior caso — a tela abre antes e se preenche quando chega.
   */
  async function fetchProvisionHints(deviceId: number): Promise<ProvisionHintsResponse | null> {
    try {
      return await apiService.get<ProvisionHintsResponse>(
        `/logs/devices/${deviceId}/provision-hints`
      )
    } catch {
      // Palpite é conveniência: sem ele a tela ainda funciona, só exige que o
      // operador preencha à mão.
      return null
    }
  }

  /**
   * Manda o servidor entrar no equipamento e configurar o envio de syslog.
   *
   * A credencial vai no corpo e não é guardada em lugar nenhum — nem aqui, nem
   * no backend. O chamador é responsável por descartá-la depois.
   */
  async function provisionDevice(
    deviceId: number,
    entrada: {
      protocol: string
      port: number | null
      username: string
      password: string
      /** Id do catálogo de sistemas — ver `stores/operatingSystems`. */
      operatingSystem: string | null
      /**
       * Endereço deste servidor que o **equipamento** deve usar. Vem do campo
       * da tela, não do navegador: quem abre a interface em `localhost` mandava
       * `localhost` para dentro do roteador, e o aparelho passava a enviar o
       * log para si mesmo — sem erro e sem nada chegando aqui.
       */
      serverAddress: string | null
      macAddress: string | null
    }
  ): Promise<ProvisionLoggingResponse> {
    const resposta = await apiService.post<ProvisionLoggingResponse>(
      `/logs/devices/${deviceId}/provision`,
      entrada
    )
    // O que acabou de ser configurado muda a lista de origens e pode ter
    // vinculado o dispositivo — recarregar aqui evita a tela mostrar o estado
    // anterior logo depois de dizer "pronto".
    await fetchSources()
    list.reset()
    return resposta
  }

  // --- guia de configuração -------------------------------------------------

  const setupGuide = ref<SetupGuide | null>(null)

  /**
   * Gera os comandos com um endereço estampado.
   *
   * O endereço vem da lista "Endereços deste servidor", não do navegador. A
   * versão anterior mandava `window.location.hostname`, e quem abre a interface
   * em `http://localhost:3333` copiava um comando com `remote=localhost` — que
   * o roteador aceita e que o faz mandar o log para si mesmo.
   */
  async function fetchSetupGuide(address: string | null = null): Promise<void> {
    try {
      const query = address ? `?address=${encodeURIComponent(address)}` : ''
      setupGuide.value = await apiService.get<SetupGuide>(`/logs/setup-snippet${query}`)
    } catch {
      setupGuide.value = null
    }
  }

  return {
    filters,
    entries: list.items,
    scrollKey: list.scrollKey,
    window: list.window,
    error: list.error,
    isEmpty,
    load: list.load,
    reset: list.reset,
    prepend: list.prepend,
    applyFilters,
    clearFilters,
    tailing,
    startTail,
    stopTail,
    toggleTail,
    sources,
    unknownCount,
    sourcesLoaded,
    nat,
    natMasking,
    fetchSources,
    bindSource,
    fetchProvisionHints,
    provisionDevice,
    setupGuide,
    fetchSetupGuide,
  }
})

/**
 * Cor da severidade do syslog no tema do Vuetify.
 *
 * Mora aqui, e não no componente, porque a aba de logs do dispositivo (Fase 4)
 * mostra a mesma tabela: duas tabelas de cor divergiriam na primeira alteração.
 */
export function severityColor(severity: number | null): string {
  if (severity === null) return 'grey'
  if (severity <= 2) return 'error'
  if (severity === 3) return 'error'
  if (severity === 4) return 'warning'
  if (severity <= 6) return 'info'
  return 'grey'
}

File: backend/src/dtos/logs.rs
//! Contrato HTTP da tela de logs.
//!
//! **O envelope não é o `{data, meta}` paginado por número de página.** O
//! `useInfiniteList` decide o fim da lista por `meta.currentPage >=
//! meta.lastPage`, e um cursor não tem `lastPage` — nem poderia ter, porque
//! contar as linhas da janela custaria um `COUNT(*)` a cada rolagem sobre
//! milhões de registros. Fabricar número de página falso para caber no
//! composable existente quebraria na primeira linha inserida durante a
//! rolagem. Daí o envelope próprio e o `useInfiniteCursor.ts` no frontend.

use serde::{Deserialize, Serialize};
use ts_rs::TS;

/// Filtros de `GET /api/logs`, como chegam na query string.
///
/// Tudo opcional: a tela abre sem filtro nenhum e recebe as últimas 24 h.
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LogsQuery {
    pub device_id: Option<i64>,
    /// Severidade numérica **máxima**. No syslog o número menor é o mais
    /// grave, então `3` significa "erro e acima" — erro, crítico, alerta e
    /// emergência.
    pub severity: Option<i16>,
    pub facility: Option<i16>,
    /// Início da janela, em RFC 3339. Ausente, valem as últimas 24 h.
    pub from: Option<String>,
    pub to: Option<String>,
    /// Busca literal na mensagem. `%` e `_` do usuário são escapados.
    pub q: Option<String>,
    /// Cursor opaco devolvido em `meta.nextCursor`.
    pub cursor: Option<String>,
    pub limit: Option<u64>,
}

/// Uma linha de log como a tela a consome.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct LogEntry {
    #[ts(type = "number")]
    pub id: i64,
    #[ts(type = "number | null")]
    pub device_id: Option<i64>,
    /// Nome do dispositivo, hidratado do banco principal. `null` quando a
    /// origem caiu numa rede cadastrada sem dispositivo, quando o vínculo é
    /// ambíguo, ou quando o dispositivo foi apagado depois — não há FK entre
    /// os dois bancos, então a linha sobrevive ao aparelho.
    pub device_name: Option<String>,
    pub source_ip: String,
    /// A verdade: quando **este** servidor recebeu. RFC 3339.
    pub received_at: String,
    /// O que o dispositivo alegou. Pode faltar, e pode estar errado — relógio
    /// de roteador sem NTP manda 1970.
    pub device_time: Option<String>,
    #[ts(type = "number | null")]
    pub facility: Option<i16>,
    #[ts(type = "number | null")]
    pub severity: Option<i16>,
    /// Rótulo pronto da severidade (`erro`, `aviso`, …), para a tela não
    /// reimplementar a tabela do RFC 5424.
    pub severity_label: Option<String>,
    pub hostname: Option<String>,
    pub app_name: Option<String>,
    #[ts(type = "number | null")]
    pub pid: Option<i32>,
    /// Tópicos do RouterOS, já quebrados em lista (`system`, `info`, …).
    pub topics: Vec<String>,
    pub message: String,
}

/// O `meta` do envelope por cursor.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct LogPageMeta {
    /// Cursor da próxima página. `null` quando acabou — é este campo, e não uma
    /// contagem, que encerra a rolagem infinita.
    pub next_cursor: Option<String>,
    pub has_more: bool,
    #[ts(type = "number")]
    pub limit: u64,
    /// A janela efetivamente consultada, depois de aplicados o padrão de 24 h e
    /// o teto de 7 dias. A tela mostra isto para o usuário não achar que está
    /// vendo o período inteiro que pediu.
    pub from: String,
    pub to: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct LogPageResponse {
    pub data: Vec<LogEntry>,
    pub meta: LogPageMeta,
}

/// Uma origem vista pelo servidor desde o último reinício.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct LogSourceEntry {
    pub source_ip: String,
    /// A chave a devolver no `bind` — o IP, ou `host:<hostname>` quando a
    /// origem chega mascarada por NAT. A tela não monta esta chave sozinha: só
    /// o servidor sabe se o endereço é gateway.
    pub bind_key: String,
    /// Se o `sourceIp` é o gateway de um NAT em vez do remetente real. Quando
    /// verdadeiro, o endereço é o mesmo para todos os equipamentos e só o
    /// hostname os separa.
    pub masked: bool,
    /// `device` | `network` | `ambiguous` | `unknown`.
    pub kind: String,
    #[ts(type = "number | null")]
    pub device_id: Option<i64>,
    pub device_name: Option<String>,
    /// Candidatos quando o mesmo IP existe em mais de um dispositivo. É o que a
    /// tela oferece no vínculo manual.
    #[ts(type = "Array<{ id: number, name: string }>")]
    pub candidates: Vec<LogSourceCandidate>,
    pub hostname: Option<String>,
    #[ts(type = "number")]
    pub message_count: u64,
    /// Quantas foram descartadas por não resolver para nada cadastrado.
    #[ts(type = "number")]
    pub dropped_count: u64,
    pub first_seen_at: String,
    pub last_seen_at: String,
    pub last_message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct LogSourceCandidate {
    #[ts(type = "number")]
    pub id: i64,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct LogSourcesResponse {
    pub data: Vec<LogSourceEntry>,
    /// Quantas origens estão descartando mensagem agora. É o número do banner
    /// da tela — sem ele, a regra "fonte desconhecida não grava" vira um buraco
    /// negro de suporte.
    #[ts(type = "number")]
    pub unknown_count: usize,
    /// Contadores de ingestão desde o boot.
    pub metrics: LogIngestMetrics,
    /// Diagnóstico do NAT. Vem sempre, para a tela poder explicar um parque
    /// inteiro "desconhecido" sem o operador ter de adivinhar.
    pub nat: LogNatDiagnostics,
}

/// O que o servidor sabe sobre o mascaramento da origem.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct LogNatDiagnostics {
    /// Quantas origens chegam com o endereço reescrito.
    #[ts(type = "number")]
    pub masked_count: usize,
    /// Dessas, quantas ainda **não** estão vinculadas a um dispositivo.
    ///
    /// É este número, e não o de cima, que decide se o aviso aparece. O
    /// mascaramento continua existindo depois de tudo vinculado — mas aí ele
    /// deixou de atrapalhar, e um aviso permanente sobre um problema resolvido
    /// treina o operador a ignorar avisos.
    #[ts(type = "number")]
    pub unresolved_masked_count: usize,
    /// Se o processo está dentro de um container — muda o texto da orientação,
    /// não o comportamento.
    pub containerized: bool,
    /// Os endereços reconhecidos como gateway, para o aviso poder nomeá-los.
    pub gateways: Vec<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct LogIngestMetrics {
    #[ts(type = "number")]
    pub received: u64,
    #[ts(type = "number")]
    pub stored: u64,
    #[ts(type = "number")]
    pub dropped_queue_full: u64,
    #[ts(type = "number")]
    pub dropped_rate_limit: u64,
    #[ts(type = "number")]
    pub dropped_unknown_source: u64,
    #[ts(type = "number")]
    pub dropped_oversized: u64,
}

/// Corpo de `POST /api/logs/sources/{ip}/bind`.
///
/// `deviceId` nulo **desfaz** o vínculo — é o mesmo endpoint, e não um `DELETE`
/// separado, porque a tela oferece um seletor onde "nenhum" é uma opção.
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BindSourceInput {
    pub device_id: Option<i64>,
}

/// Corpo de `POST /api/logs/devices/{id}/provision`.
///
/// **Estes campos não são persistidos.** `username` e `password` existem
/// enquanto a requisição dura e morrem com ela — não há coluna, cache nem
/// `system_settings` que os receba, nem cifrados. Ver a nota do módulo
/// [`crate::services::syslog::provision`]. O DTO fica sem `Debug` derivado de
/// propósito: um `tracing` distraído despejaria a senha no log.
#[derive(Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProvisionLoggingInput {
    /// `ssh`, `telnet` ou `mactelnet`.
    pub protocol: String,
    /// Ausente cai na porta padrão do protocolo (22, 23 ou 20561).
    pub port: Option<u16>,
    pub username: String,
    pub password: String,
    /// Id do catálogo de sistemas. Ausente, o servidor deduz do cadastro.
    pub operating_system: Option<String>,
    /// Endereço deste servidor como o **equipamento** o alcança. Vazio, ou
    /// `localhost`, faz o servidor descobrir sozinho: gravar `localhost` no
    /// roteador o faria mandar o log para si mesmo, sem erro e sem nada
    /// chegando aqui.
    pub server_address: Option<String>,
    /// MAC do equipamento, para o MAC-Telnet. Ausente, o servidor procura nas
    /// interfaces coletadas e no resultado da descoberta.
    pub mac_address: Option<String>,
}

/// O que a tela usa para se preencher antes de perguntar qualquer coisa.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct ProvisionHintsResponse {
    /// Endereço deste servidor que o equipamento deve usar. `null` quando não
    /// há resposta confiável — aí a tela exige que o operador digite.
    pub server_address: Option<String>,
    /// De onde veio o palpite, para a tela não apresentar chute como certeza.
    pub server_address_source: String,
    /// Qual entrada da lista de endereços do servidor serve a **este**
    /// equipamento. Nulo quando não há evidência para escolher — a tela então
    /// pergunta ao operador em vez de chutar um endereço.
    pub suggested_address_id: Option<String>,
    /// Por que aquela foi sugerida — "este servidor alcança o equipamento por
    /// 10.8.0.1". O motivo é o que dispensa o operador de entender a lista.
    pub suggested_address_reason: Option<String>,
    /// Forma de acesso em vigor: `local`, `vpn` ou `remote`.
    pub access_mode: String,
    /// Se ela veio do cadastro. A tela diz "definido no cadastro" em vez de
    /// "deduzido", e as duas frases pedem confiança diferente do operador.
    pub access_mode_declared: bool,
    /// Por que esta forma de acesso.
    pub access_mode_reason: String,
    #[ts(type = "number")]
    pub server_port: u16,
    /// Sistema em vigor — id do catálogo de `services::devices::systems`.
    pub operating_system: String,
    /// De onde ele veio: `declarado`, `snmp`, `sonda`, `cadastro` ou `padrão`.
    pub operating_system_source: String,
    /// A frase que explica a conclusão — "o servidor SSH se identifica como
    /// `dropbear`, que é o padrão do OpenWrt". Sem ela a tela só afirma.
    pub operating_system_reason: String,
    /// Porta 22 respondeu à sondagem.
    pub ssh_open: bool,
    /// Porta 23 respondeu à sondagem.
    pub telnet_open: bool,
    pub mac_address: Option<String>,
    /// Se este processo alcança a camada 2 da rede do equipamento. Falso num
    /// container em rede bridge, onde o MAC-Telnet não tem como funcionar.
    pub layer2_reachable: bool,
}

/// O que a tela mostra depois da ativação automática.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct ProvisionLoggingResponse {
    pub operating_system: String,
    /// O endereço que foi gravado no equipamento.
    pub server_address: String,
    #[ts(type = "number")]
    pub server_port: u16,
    /// Os comandos enviados. Nunca a credencial.
    pub commands: Vec<String>,
    /// O que o equipamento respondeu, com a senha raspada.
    pub transcript: String,
    /// Se chegou log do dispositivo antes do teto de espera. `null` quando não
    /// havia como confirmar (ingestão desligada).
    pub confirmed: Option<bool>,
}

/// Resposta do vínculo/desvínculo manual de uma origem de syslog.
#[derive(Debug, Clone, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export, export_to = "../../frontend/src/bindings/")]
pub struct BindSourceResponse {
    pub bind_key: String,
    #[ts(type = "number | null")]
    pub device_id: Option<i64>,
}

/// Filtros do live tail (`GET /api/logs/stream`).
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LogStreamQuery {
    pub device_id: Option<i64>,
    pub severity: Option<i16>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bind_source_response_serializa_em_camel_case() {
        let response = BindSourceResponse {
            bind_key: "192.168.1.1".into(),
            device_id: Some(7),
        };
        let json = serde_json::to_value(&response).unwrap();
        assert_eq!(json["bindKey"], "192.168.1.1");
        assert_eq!(json["deviceId"], 7);
    }
}
